### ISSUE TITLE
[tbb] Please support tbb version 2021.1.1 

### ISSUE THREAD

**Issue by JackBoosY** [PixarAnimationStudios/USD#1471]
Hi guys, I'm vcpkg maintainer.
I tried to update tbb to the latest version 2021.1.1 in vcpkg (PR microsoft/vcpkg#16321), but seems that usd doesn't support this version. And some other libraries has been adapted for that.
Can you consider updating the code to adapt to this version?

Thanks.

**Comment by jilliene** [comment]
Filed as internal issue #USD-6600

**Comment by manuelk** [comment]
Confirmed - after much trial & error, was able to get a working build using the 'tbb_2019' branch (which unfortunately did not have a cmake build or exported config yet).

The biggest hurdle seems to be the obsoleting of tbb::atomic (replaced with std::atomic sometime in late 2020). 

IMHO, there is little reason to not submodule tbb & just let tbb's cmake config take care of everything, but just in case, I am attaching the snippet i had to modify in the FindTBB.cmake module (because the version file moved too)

```
  find_file(TBB_VERSION_FILE 
      NAMES oneapi/tbb/version.h tbb/version.h tbb/tbb_stddef.h       
      HINTS ${TBB_INCLUDE_DIRS} ${TBB_SEARCH_DIR}
      PATHS ${TBB_DEFAULT_SEARCH_DIR})

  if(TBB_VERSION_FILE)
    file(READ "${TBB_VERSION_FILE}" _tbb_version_file)

    string(REGEX REPLACE ".*#define TBB_VERSION_MAJOR ([0-9]+).*" "\\1"
        TBB_VERSION_MAJOR "${_tbb_version_file}")

    string(REGEX REPLACE ".*#define TBB_VERSION_MINOR ([0-9]+).*" "\\1"
        TBB_VERSION_MINOR "${_tbb_version_file}")

    string(REGEX REPLACE ".*#define TBB_INTERFACE_VERSION ([0-9]+).*" "\\1"
        TBB_INTERFACE_VERSION "${_tbb_version_file}")

    set(TBB_VERSION "${TBB_VERSION_MAJOR}.${TBB_VERSION_MINOR}")
  endif()
```

**Comment by meshula** [comment]
Maybe the biggest hurdle can be cleared by moving USD to std::atomic :) You know what gets my vote! @manuelk , is that the approach you took?

**Comment by manuelk** [comment]
the compiler was throwing lots of errors, so i didn't really try to fix it (with the idea that other things may have been obsoleted and apparently USD was not being kept up to date obviously)

took the easy out & just reverted TBB to a 2019 version

i do agree with you that a lot of the 'arch' code in USD should be moved to modern C++ (mutex, "ref" ptrs, ...) with the hope of maybe killing the now extremely unwieldy TBB dependency (full win10 binary installer is 22GB on disk - for real Intel !)

**Comment by manuelk** [comment]
Something else that makes no sense w/ TBB handling in USD :

in the TBB module you correctly define a debug & release interface (assuming the expression works - it's hard to read...)
```
      set_target_properties(TBB::tbb PROPERTIES
          INTERFACE_COMPILE_DEFINITIONS "$<$<OR:$<CONFIG:Debug>,$<CONFIG:RelWithDebInfo>>:TBB_USE_DEBUG=1>"
          IMPORTED_LOCATION_DEBUG          ${TBB_LIBRARIES_DEBUG}
          IMPORTED_LOCATION_RELWITHDEBINFO ${TBB_LIBRARIES_DEBUG}
          IMPORTED_LOCATION_RELEASE        ${TBB_LIBRARIES_RELEASE}
          IMPORTED_LOCATION_MINSIZEREL     ${TBB_LIBRARIES_RELEASE}
          )
```

but then all the USD components bypass the interface to link directly with the component variable:
```
pxr_library(tf
    LIBRARIES
        arch
        ${WINLIBS}
        ${PYTHON_LIBRARIES}
        ${Boost_PYTHON_LIBRARY}
        ${TBB_tbb_LIBRARY}
```

Obviously this does not work well for debug builds, as neither TBB_USE_DEBUG is set, nor is the linker pointed to the right library.

**Comment by e4lam** [comment]
@jilliene @meshula We're moving up to TBB 2020 update 3 for the next major release of Houdini right now (since CY2021 vfxplatform.com is TBB 2020). I'm noticing the same issues except they're flagged as deprecation warnings instead of out right removals. I think it'll be worthwhile to first move USD up to TBB 2020 and replace uses as we go to remove the deprecated functionality.

**Comment by meshula** [comment]
@manuelk Yeah, it would be more correct/bulletproof to link via TBB::tbb, not ${TBB_tbb_LIBRARY} (as elaborated in the plethora of "modern cmake" articles.)

**Comment by HnimNart** [comment]
Do you have a timeline for updating to TBB 2021?

**Comment by c64kernal** [comment]
We don't unfortunately have a timeline for moving to newer versions of tbb... though we typically try to at least follow http://vfxplatform.com/

**Comment by manuelk** [comment]
I've just run a quick scan for TBB over the source code - the majority of the usage seems to be for things like mutex, atomic and thread local storage, which should be transitioned to modern C++ (as recommended by Intel in TBB).

However, there also seems to be enough use of thread arenas, concurrent queues, etc that deprecating TBB would not be a small task (it's really unfortunate that OneTBB has become so unwieldy in recent years). But it's still nothing compared to the headaches of boost + python though.

**Comment by spiffmon** [comment]
@manuelk , even if somehow there were enough resources in the universe (so to speak) to wean USD off of TBB, that in itself would create an even less tractable problem:  for better or worse, TBB has becomes the CPU resource management system of the VFX industry, so if USD were to move away from it, it would become really difficult to share resources with host applications in a good-citizen **and** performant way.

I think the only realistic way to remove the TBB dependency from USD is if the VFX platform itself mandates it, so that USD and all its consumers could move together.

**Comment by manuelk** [comment]
There be a boat anchor... guess we can commiserate around beers at the next non-virtual Siggraph :) (if ever)

**Comment by meshula** [comment]
Spiff points out that using TBB helps alleviate CPU over-subscription, a might important consideration. That does leave some room for some targeted weaning, around some std provided constructs like atomics, and particularly around the parts of TBB slated for removal, at least on vfxplatform gated time table.

Sidebar: To get the same benefit on the over-subscription issue on Apple platforms, the Work system would need to cooperate with Grand Central Dispatch. Perhaps longer term, there's more to discuss around that, as an independent topic.

**Comment by ggris** [comment]
Hi, guys
I ended up here because I wanted to to update TBB through VCPKG in our codebase and followed the threads to see what was the issue and how I could maybe help solving it.

I'd like to share why we have to update it or our codebase because after looking at this repo, it might have implications on your side. It all boils down to spin lock and why unless you have a really guru expert, you should not play with fire and try to save some performance by using them instead of regular mutexes. 

Long story short: in their internal task_queue, prior to the 2021 release, they were using a spin lock to protect the queue which, as is often the case with spin lock, resulted on occasional lock access freeze when trying to access the queue.

This is probably relevant to this repo because I see `tbb::spin_mutex` used all over the place. So either
 - Your lib is never used on high CPU loads or on single core CPU
 - You have a real threading expert internally who understands the issue I'm talking about and reviewed all spin locks usage to guarantee none of them are problematic (But note that even the guys at tbb did mistakes when using spin locks)
 - This is all legacy code and no one really knows why it's spin mutex and not regular mutex

**Comment by spiffmon** [comment]
Thanks @ggris !  We monitor our performance suite for lock contention, and have deployed several strategies for reducing it when it crops up.  We have not observed correctness issues or deadlocks attributable to our use of tbb's spin locks, but if you have evidence of such a problem, or of scalability issues (we do not regularly run on more than 32 cores yet), please let us know.

**Comment by e4lam** [comment]
@ggris On a side note, can you point to where the fix is for TBB? Thanks!

**Comment by ggris** [comment]
@e4lam Actually, I was mistaken in thinking it was already merged but it is a fix for this ticket: https://github.com/oneapi-src/oneTBB/issues/546
 Note that in this case the issue comes from an implicit spin lock pattern. As I learned while researching this, it seems TBB spin_mutex is protected against the typical spin lock pattern issue with a backoff strategy and the solution for this ticket is to apply similar backoff strategy in this implicit spin lock pattern. I've had so many issues in the past with libraries using bad implementation of the spin lock pattern that I now assume by default that all spin lock patterns are buggy.

I ended up backporting the backoff fix to our internal build of TBB as it is unlikely we'll get an up to date TBB

**Comment by mbechard** [comment]
@c64kernal at the very least could the codebase remove usage of deprecated (now removed) TBB features and move to the std:: replacements? That way newer TBB binaries can be used at runtime. Even if USD is built against TBB 2020 Update 3, if it's not using features that have been removed in 2021.4.0, we should be able to drop in 2021.4.0 binaries (which other tools we use may rely on) and allow things to still function.
An upgrade guide can be found here:
https://www.intel.com/content/www/us/en/developer/articles/technical/tbb-revamp.html

**Comment by c64kernal** [comment]
@mbechard -- yes for sure! If there is a minimally invasive change that you'd like to PR, we'd love to have a look and would definitely consider. We're currently constrained to C++14 by the way. Thanks so much for pushing through this!

**Comment by mbechard** [comment]
@c64kernal unfortunately I don't think I'm in a position to do a PR for this. Certainly I could upgrade the tbb::atomics, but things such as tbb::empty_task and related threading work would likely be tricky for me to do as I've spent no time working in the USD codebase (and really, barely any in TBB as well).
Since these changes are compatible with both the current codebase and in the future (and will be required once vfxplatforms reaches TBB 2021), I was suggesting they get added sooner rather than later to allow people to use the newer TBB when required. 
Thanks for your consideration!

**Comment by c64kernal** [comment]
Okay got it, thanks again @mbechard -- I filed an internal issue to do this initial clean up, and we'll prioritize as best we can. Thanks again for the suggestion.

**Comment by sybrenstuvel** [comment]
Hello,

A little introduction: I'm Sybren, the developer who added USD support to [Blender](https://www.blender.org/).

> We don't unfortunately have a timeline for moving to newer versions of tbb... though we typically try to at least follow http://vfxplatform.com/

This puts anyone who wants to use USD in an application in a tough spot. The VFX Reference Platform has a strange mix of super new versions of libraries and deprecated/outdated versions. For example, for a while 'CY2021' targeted a not-yet-released version (3.0) of OpenEXR while targeting a version of Python (3.7) that was so old it would be unsupported for the 2nd half of the year.

If USD developers follow VFX Reference Platform in the sense of only supporting deprecated versions of these libraries, this means that any application that wants to use USD also has to lag behind.

USD is using parts of the TBB API that were deprecated two years ago. Various Linux distributions will ship with a newer version soon; they typically only allow installing a single version of such libraries, which means that building Blender with USD support on contemporary Linuxes is going to be problematic.

> I filed an internal issue to do this initial clean up, and we'll prioritize as best we can.

This was posted about 4 months ago. What is the current status?

PS: OpenSubDiv also still only supports the deprecated TBB API, but there at least [patching the library is fairly trivial](https://github.com/PixarAnimationStudios/OpenSubdiv/pull/1234). For USD the work seems to be more considerable, though, which means that Blender can't just include a patch to make it work.

**Comment by spiffmon** [comment]
Hi @sybrenstuvel ,
    We understand your vexation, and honestly would very much like to be on TBB's OneAPI.  We _hope_ to be able to do this upgrade by end of year. 
 To understand why it is taking us so long, there are two factors at play, which are related: resources, and sensitivity to the VFX industry.

You may think the VFX reference platform is criminally impeding forward progress, but in fact it protects animation and vfx studios from vendor thrashing, given our (unfortunate, but difficult to change) slow pace of updating OS's and system components.  I am attaching a fascinating [survey of 88 anim/Vfx studios conducted last year by the VES](https://github.com/PixarAnimationStudios/USD/files/8249262/2021.Studio.Platform.Survey.Report.pdf), for your information.  It shows where the industry is in terms of OS's, but the thing I really want to highlight to you is that until the (Predicted) end of this year, not even a **majority** of VFX houses will have migrated to Python 3.x yet!  Pixar is, in fact, in the middle of wrestling our large internal codebase into 3.x compatibility, and even though (thanks to NVidia!) USD _can_ be built with python 3, internally all production and development builds still use python 2.7x.

Referring back to the reference platform, then, we and other studios are still trying to adopt CY2020, and the move to OneAPI does not happen until CY2023.  We _may_ jump right to CY2022 after finishing 2020 adoption, but in either case, even CY2022 is only TBB 2020, reflective of the widespread expectation that moving forward to OneApi is going to be a Big Deal.  In fact, we do not yet even know how we are going to manage it without serious performance degradation to some of our internal software, because OneApi removes some lower-level API's that we leverage to good effect.

If you or others have the resources and will to want to see USD be OneAPI compatible ahead of our ability to make it so, we would be very appreciative.  The constraints, as we see them, are:
1. Anything that would want to land soon would, in deference to vfx studios and the software they are currently using,  need to support building against **either** TBB 2019 _or_ 202x .
2. As you mentioned, USD still uses some TBB containers/locks that have been deprecated in favor of std options.  But many of those are in performance-sensitive codesites, so adopting alternatives _may_ require extra consideration.  We are working, for other collaborations, to make available some of the performance tests and test assets we rely on internally, but that may not bear fruit soon enough to be useful.

That said, I do think such a contribution would help speed this process up.

Thanks, Sybren!

**Comment by boberfly** [comment]
Hi all,

Just a heads-up I started to have a crack at it today and have made some progress: 
https://gist.github.com/boberfly/07f7f8a4c7433d8ebd5cdfa640ef7758

As you can see in the error log, it still has an issue and my head hurts from all the template+namespace+function bindings... :)

https://oneapi-src.github.io/oneTBB/main/tbb_userguide/Migration_Guide/Task_API.html

Cheers

**Comment by e4lam** [comment]
@boberfly I'd suggest that you split it out under separate PR's by porting in different branches if possible, one for each category of API changes. In particular, I think at least the atomic changes should be separate.

**Comment by boberfly** [comment]
@e4lam good idea, I am just trying to get this working on regular TBB 2020U3 to verify my `TBB_INTERFACE_VERSION` defines and `std::atomic` replacements are working, unfortunately this one in particular: https://github.com/PixarAnimationStudios/USD/blob/release/pxr/usd/usdGeom/bboxCache.cpp#L107 causes template deduce issues for some reason on [this](https://github.com/PixarAnimationStudios/USD/blob/release/pxr/usd/usdGeom/bboxCache.cpp#L153) line for the insert which seems bonkers to me at the moment, if anyone else has any ideas... :)

**Comment by furby-tm** [comment]
@boberfly The bboxCache has been the bane of my existence in terms of being able to upgrade TBB. Every time I've went through all the work you've done - I have gotten stuck at this exact section. If you know how to fix it that would be huge.

We may need to reach out to the Intel TBB team to get this resolved, I've been staring at this issue for years 😆.

This issue also exists on all operating systems - macOS, Windows, and Linux.

**Comment by furby-tm** [comment]
@sybrenstuvel I'd be happy to actively maintain a custom USD build that fits Blender's dependency needs, as it is something I typically find myself doing anyway.

**Comment by meshula** [comment]
@boberfly, it could be the compiler is having trouble with the make_pair rvalue with the later version of tbb. A suggestion - the way I'd go about debugging is to break it out onto its own line as a lvalue, either as an auto or an explicitly declared pair.

**Comment by sybrenstuvel** [comment]
> @sybrenstuvel I'd be happy to actively maintain a custom USD build that fits Blender's dependency needs, as it is something I typically find myself doing anyway.

That is a very generous offer @furby-tm . Please pop over to the [#blender-builds channel](https://blender.chat/channel/blender-builds) on Blender Chat to talk about this. That way the other platform maintainers (and other interested parties) are also in the loop.

**Comment by boberfly** [comment]
Hey all,
Thanks for the hint @meshula - I believe I've found the root cause of the issue - the move constructor is undefined for `std::atomic` - so the `std::make_pair` isn't going to work. An `emplace` could be used but I think this isn't implemented yet in `TfHashMap` from `<unordered_map>`.

**Comment by meshula** [comment]
@boberfly So the issue is that _PrototypeTask can't be moved because of the contained  tbb::atomic (I guess you are replacing it with std::atomic)? Perhaps adding move constructor/assignment to _PrototypeTask would do the job. Since the intent is to create a new _PrototypeTask at that point in the code, there wouldn't yet be an opportunity for contention on the atomic value.

**Comment by ggris** [comment]
If it is <unordered_map> based, you may replace the `insert` with `try_emplace` and it should work.  https://godbolt.org/z/3P93Yjcq9
If it is not @meshula solution is probably reasonable

**Comment by boberfly** [comment]
@ggris Sadly just enhancing TfHashMap isn't enough for Linux+GCC:
```
#if defined(ARCH_HAS_GNU_STL_EXTENSIONS)
#include <ext/hash_map>
```
This implementation doesn't come with `emplace/try_emplace` but if I just use `std::unordered_map` as the `_PrototypeTaskMap` typedef for this, it's working!

However, taking @meshula 's advice and just filling in copy/move constructors also solves things. As well as another one `_Entry` in `resolvedAttributeCache.h`

With the help of @furby-tm and @deadpin (Jesse Y) we've come to this branch so far:
https://github.com/boberfly/usd/tree/std_atomic

One part I'm not fully confident with is the `compare_and_swap` replacement which will need additional eyes, then I think this is ready for a PR.

**Comment by meshula** [comment]
Refering to your branch and searching for the compare_and_swap, If you are asking about using the _weak variant, I think maybe you want _strong. If _weak fails spuriously, it will result in hitting a hard spinlock in the following lines of code. _strong will probably have better behavior, in that it won't spuriously fail, and thus spuriously spin. @gitamohr would be the person I'd want to ask about this one.

**Comment by gitamohr** [comment]
I'm not an expert on that imaging code specifically, but yeah generally if you're doing a CAS with retries until it succeeds (i.e. in a loop) then using the _weak version is usually okay.  But if you're not retries or if for any other reason a spurious failure would be a problem then use the _strong.  This case looks like a _strong candidate, and since you're replacing tbb::atomic (which AFAIK always has _strong semantics) I'd go with that.

**Comment by ggris** [comment]
As a side note, this is one of the places where USD should use a mitigation strategy for the spin loop (as the comment rightfully points out in the code). TBB has the nice `tbb::internal::atomic_backoff` for this kind of situation. Though, as the namespace suggests, it is only an internal utility.
Though it is probably not the right PR to address this issue

**Comment by e4lam** [comment]
@boberfly Looks pretty good, just a minor comment that I added to the commit:
https://github.com/boberfly/USD/commit/e0acaa1bf7a279650a58627cc33554ea15a391db#r76711806

We removed a whole bunch of `<tbb/atomic.h>` includes but I feel like we need to add back some `<atomic>` includes. I wouldn't necessarily add them back to the PCH but perhaps at least to the header files that have classes which contain `std::atomic` data members.

**Comment by jiapei100** [comment]
Yeah.... Looks the current problems focus on 2 files: [dispatcher.h](https://github.com/PixarAnimationStudios/USD/blob/release/pxr/base/work/dispatcher.h) and [dispatcher.cpp](https://github.com/PixarAnimationStudios/USD/blob/release/pxr/base/work/dispatcher.cpp) .



```console
[ 43%] Built target _tf
....../USD/pxr/base/work/threadLimits.cpp:63:13: error: ‘task_scheduler_init’ in namespace ‘tbb’ does not name a type
   63 | static tbb::task_scheduler_init *_tbbTaskSchedInit;
      |             ^~~~~~~~~~~~~~~~~~~
....../USD/pxr/base/work/threadLimits.cpp: In function ‘unsigned int pxrInternal_v0_22__pxrReserved__::WorkGetPhysicalConcurrencyLimit()’:
....../USD/pxr/base/work/threadLimits.cpp:70:17: error: ‘tbb::task_scheduler_init’ has not been declared
   70 |     return tbb::task_scheduler_init::default_num_threads();
      |                 ^~~~~~~~~~~~~~~~~~~
....../USD/pxr/base/work/threadLimits.cpp: In function ‘void pxrInternal_v0_22__pxrReserved__::Work_InitializeThreading()’:
....../USD/pxr/base/work/threadLimits.cpp:126:9: error: ‘_tbbTaskSchedInit’ was not declared in this scope
  126 |         _tbbTaskSchedInit = new tbb::task_scheduler_init(threadLimit);
      |         ^~~~~~~~~~~~~~~~~
....../USD/pxr/base/work/threadLimits.cpp:126:33: error: expected type-specifier
  126 |         _tbbTaskSchedInit = new tbb::task_scheduler_init(threadLimit);
      |                                 ^~~
....../USD/pxr/base/work/threadLimits.cpp:117:14: warning: unused variable ‘threadLimit’ [-Wunused-variable]
  117 |     unsigned threadLimit =
      |              ^~~~~~~~~~~
....../USD/pxr/base/work/threadLimits.cpp: In function ‘void pxrInternal_v0_22__pxrReserved__::WorkSetConcurrencyLimit(unsigned int)’:
....../USD/pxr/base/work/threadLimits.cpp:165:9: error: ‘_tbbTaskSchedInit’ was not declared in this scope
  165 |     if (_tbbTaskSchedInit) {
      |         ^~~~~~~~~~~~~~~~~
....../USD/pxr/base/work/threadLimits.cpp:169:33: error: expected type-specifier
  169 |         _tbbTaskSchedInit = new tbb::task_scheduler_init(threadLimit);
      |                                 ^~~
make[2]: Leaving directory '....../USD/build'
make  -f pxr/base/vt/CMakeFiles/vt.dir/build.make pxr/base/vt/CMakeFiles/vt.dir/build
make[2]: Entering directory '....../USD/build'
make[2]: Nothing to be done for 'pxr/base/vt/CMakeFiles/vt.dir/build'.
make[2]: Leaving directory '....../USD/build'
make[2]: *** [pxr/base/work/CMakeFiles/work.dir/build.make:208: pxr/base/work/CMakeFiles/work.dir/threadLimits.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 44%] Built target vt
make  -f pxr/base/vt/CMakeFiles/_vt.dir/build.make pxr/base/vt/CMakeFiles/_vt.dir/depend
make  -f pxr/imaging/pxOsd/CMakeFiles/pxOsd.dir/build.make pxr/imaging/pxOsd/CMakeFiles/pxOsd.dir/depend
make  -f pxr/base/vt/CMakeFiles/testVtCpp.dir/build.make pxr/base/vt/CMakeFiles/testVtCpp.dir/depend
make[2]: Entering directory '....../USD/build'
cd ....../USD/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" ....../USD ....../USD/pxr/base/vt ....../USD/build ....../USD/build/pxr/base/vt ....../USD/build/pxr/base/vt/CMakeFiles/_vt.dir/DependInfo.cmake --color=
make[2]: Entering directory '....../USD/build'
cd ....../USD/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" ....../USD ....../USD/pxr/base/vt ....../USD/build ....../USD/build/pxr/base/vt ....../USD/build/pxr/base/vt/CMakeFiles/testVtCpp.dir/DependInfo.cmake --color=
make[2]: Entering directory '....../USD/build'
cd ....../USD/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" ....../USD ....../USD/pxr/imaging/pxOsd ....../USD/build ....../USD/build/pxr/imaging/pxOsd ....../USD/build/pxr/imaging/pxOsd/CMakeFiles/pxOsd.dir/DependInfo.cmake --color=
Consolidate compiler generated dependencies of target testVtCpp
Consolidate compiler generated dependencies of target pxOsd
Consolidate compiler generated dependencies of target _vt
make[2]: Leaving directory '....../USD/build'
make  -f pxr/base/vt/CMakeFiles/testVtCpp.dir/build.make pxr/base/vt/CMakeFiles/testVtCpp.dir/build
make[2]: Entering directory '....../USD/build'
make[2]: Nothing to be done for 'pxr/base/vt/CMakeFiles/testVtCpp.dir/build'.
make[2]: Leaving directory '....../USD/build'
[ 44%] Built target testVtCpp
In file included from ....../USD/pxr/base/work/dispatcher.cpp:26:
....../USD/pxr/base/work/dispatcher.h:140:48: error: expected class-name before ‘{’ token
  140 |     struct _InvokerTask : public tbb::v1::task {
      |                                                ^
....../USD/pxr/base/work/dispatcher.h:147:26: error: ‘task’ in namespace ‘tbb::v1’ does not name a type
  147 |         virtual tbb::v1::task* execute() {
      |                          ^~~~
In file included from /opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/enumerable_thread_specific.h:36,
                 from /opt/intel/oneapi/tbb/latest/include/tbb/enumerable_thread_specific.h:17,
                 from ....../USD/pxr/base/tf/diagnosticMgr.h:46,
                 from ....../USD/pxr/base/tf/errorMark.h:30,
                 from ....../USD/pxr/base/work/dispatcher.h:33,
                 from ....../USD/pxr/base/work/dispatcher.cpp:26:
/opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/task.h:26:11: note: ‘tbb::v1::task’ declared here
   26 | namespace task {
      |           ^~~~
In file included from ....../USD/pxr/base/work/dispatcher.cpp:26:
....../USD/pxr/base/work/dispatcher.h:175:41: error: field ‘_context’ has incomplete type ‘tbb::detail::d1::task_group_context’
  175 |     tbb::detail::d1::task_group_context _context;
      |                                         ^~~~~~~~
In file included from /opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/task.h:22,
                 from /opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/enumerable_thread_specific.h:36,
                 from /opt/intel/oneapi/tbb/latest/include/tbb/enumerable_thread_specific.h:17,
                 from ....../USD/pxr/base/tf/diagnosticMgr.h:46,
                 from ....../USD/pxr/base/tf/errorMark.h:30,
                 from ....../USD/pxr/base/work/dispatcher.h:33,
                 from ....../USD/pxr/base/work/dispatcher.cpp:26:
/opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/detail/_task.h:44:7: note: forward declaration of ‘class tbb::detail::d1::task_group_context’
   44 | class task_group_context;
      |       ^~~~~~~~~~~~~~~~~~
In file included from ....../USD/pxr/base/work/dispatcher.cpp:26:
....../USD/pxr/base/work/dispatcher.h:176:10: error: ‘empty_task’ in namespace ‘tbb’ does not name a type
  176 |     tbb::empty_task* _rootTask;
      |          ^~~~~~~~~~
In file included from ....../USD/pxr/base/work/detachedTask.h:32,
                 from ....../USD/pxr/base/work/detachedTask.cpp:26:
....../USD/pxr/base/work/dispatcher.h:140:48: error: expected class-name before ‘{’ token
  140 |     struct _InvokerTask : public tbb::v1::task {
      |                                                ^
....../USD/pxr/base/work/dispatcher.h:147:26: error: ‘task’ in namespace ‘tbb::v1’ does not name a type
  147 |         virtual tbb::v1::task* execute() {
      |                          ^~~~
In file included from /opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/enumerable_thread_specific.h:36,
                 from /opt/intel/oneapi/tbb/latest/include/tbb/enumerable_thread_specific.h:17,
                 from ....../USD/pxr/base/tf/diagnosticMgr.h:46,
                 from ....../USD/pxr/base/tf/errorMark.h:30,
                 from ....../USD/pxr/base/work/detachedTask.h:30,
                 from ....../USD/pxr/base/work/detachedTask.cpp:26:
/opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/task.h:26:11: note: ‘tbb::v1::task’ declared here
   26 | namespace task {
      |           ^~~~
In file included from ....../USD/pxr/base/work/detachedTask.h:32,
                 from ....../USD/pxr/base/work/detachedTask.cpp:26:
....../USD/pxr/base/work/dispatcher.h:175:41: error: field ‘_context’ has incomplete type ‘tbb::detail::d1::task_group_context’
  175 |     tbb::detail::d1::task_group_context _context;
      |                                         ^~~~~~~~
In file included from /opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/task.h:22,
                 from /opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/enumerable_thread_specific.h:36,
                 from /opt/intel/oneapi/tbb/latest/include/tbb/enumerable_thread_specific.h:17,
                 from ....../USD/pxr/base/tf/diagnosticMgr.h:46,
                 from ....../USD/pxr/base/tf/errorMark.h:30,
                 from ....../USD/pxr/base/work/detachedTask.h:30,
                 from ....../USD/pxr/base/work/detachedTask.cpp:26:
/opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/detail/_task.h:44:7: note: forward declaration of ‘class tbb::detail::d1::task_group_context’
   44 | class task_group_context;
      |       ^~~~~~~~~~~~~~~~~~
In file included from ....../USD/pxr/base/work/detachedTask.h:32,
                 from ....../USD/pxr/base/work/detachedTask.cpp:26:
....../USD/pxr/base/work/dispatcher.h:176:10: error: ‘empty_task’ in namespace ‘tbb’ does not name a type
  176 |     tbb::empty_task* _rootTask;
      |          ^~~~~~~~~~
....../USD/pxr/base/work/dispatcher.h: In member function ‘void pxrInternal_v0_22__pxrReserved__::WorkDispatcher::Run(Callable&&)’:
....../USD/pxr/base/work/dispatcher.h:106:9: error: ‘_rootTask’ was not declared in this scope
  106 |         _rootTask->spawn(_MakeInvokerTask(std::forward<Callable>(c)));
      |         ^~~~~~~~~
In file included from ....../USD/pxr/base/work/detachedTask.h:32,
                 from ....../USD/pxr/base/work/utils.h:31,
                 from ....../USD/pxr/base/work/utils.cpp:26:
....../USD/pxr/base/work/dispatcher.h:140:48: error: expected class-name before ‘{’ token
  140 |     struct _InvokerTask : public tbb::v1::task {
      |                                                ^
....../USD/pxr/base/work/dispatcher.h:147:26: error: ‘task’ in namespace ‘tbb::v1’ does not name a type
  147 |         virtual tbb::v1::task* execute() {
      |                          ^~~~
In file included from /opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/enumerable_thread_specific.h:36,
                 from /opt/intel/oneapi/tbb/latest/include/tbb/enumerable_thread_specific.h:17,
                 from ....../USD/pxr/base/tf/diagnosticMgr.h:46,
                 from ....../USD/pxr/base/tf/errorMark.h:30,
                 from ....../USD/pxr/base/work/detachedTask.h:30,
                 from ....../USD/pxr/base/work/utils.h:31,
                 from ....../USD/pxr/base/work/utils.cpp:26:
/opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/task.h:26:11: note: ‘tbb::v1::task’ declared here
   26 | namespace task {
      |           ^~~~
In file included from ....../USD/pxr/base/work/detachedTask.h:32,
                 from ....../USD/pxr/base/work/utils.h:31,
                 from ....../USD/pxr/base/work/utils.cpp:26:
....../USD/pxr/base/work/dispatcher.h:175:41: error: field ‘_context’ has incomplete type ‘tbb::detail::d1::task_group_context’
  175 |     tbb::detail::d1::task_group_context _context;
      |                                         ^~~~~~~~
In file included from /opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/task.h:22,
                 from /opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/enumerable_thread_specific.h:36,
                 from /opt/intel/oneapi/tbb/latest/include/tbb/enumerable_thread_specific.h:17,
                 from ....../USD/pxr/base/tf/diagnosticMgr.h:46,
                 from ....../USD/pxr/base/tf/errorMark.h:30,
                 from ....../USD/pxr/base/work/detachedTask.h:30,
                 from ....../USD/pxr/base/work/utils.h:31,
                 from ....../USD/pxr/base/work/utils.cpp:26:
/opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/detail/_task.h:44:7: note: forward declaration of ‘class tbb::detail::d1::task_group_context’
   44 | class task_group_context;
      |       ^~~~~~~~~~~~~~~~~~
In file included from ....../USD/pxr/base/work/detachedTask.h:32,
                 from ....../USD/pxr/base/work/utils.h:31,
                 from ....../USD/pxr/base/work/utils.cpp:26:
....../USD/pxr/base/work/dispatcher.h:176:10: error: ‘empty_task’ in namespace ‘tbb’ does not name a type
  176 |     tbb::empty_task* _rootTask;
      |          ^~~~~~~~~~
....../USD/pxr/base/work/dispatcher.h: In member function ‘void pxrInternal_v0_22__pxrReserved__::WorkDispatcher::Run(Callable&&)’:
....../USD/pxr/base/work/dispatcher.h:106:9: error: ‘_rootTask’ was not declared in this scope
  106 |         _rootTask->spawn(_MakeInvokerTask(std::forward<Callable>(c)));
      |         ^~~~~~~~~
....../USD/pxr/base/work/dispatcher.h: In member function ‘pxrInternal_v0_22__pxrReserved__::WorkDispatcher::_InvokerTask<typename std::remove_reference<_Tp>::type>& pxrInternal_v0_22__pxrReserved__::WorkDispatcher::_MakeInvokerTask(Fn&&)’:
....../USD/pxr/base/work/dispatcher.h:163:22: error: ‘_rootTask’ was not declared in this scope
  163 |         return *new( _rootTask->allocate_additional_child_of(*_rootTask) )
      |                      ^~~~~~~~~
....../USD/pxr/base/work/dispatcher.cpp: In constructor ‘pxrInternal_v0_22__pxrReserved__::WorkDispatcher::WorkDispatcher()’:
....../USD/pxr/base/work/dispatcher.cpp:32:46: error: incomplete type ‘tbb::detail::d1::task_group_context’ used in nested name specifier
   32 |         tbb::detail::d1::task_group_context::isolated,
      |                                              ^~~~~~~~
....../USD/pxr/base/work/dispatcher.cpp:33:46: error: incomplete type ‘tbb::detail::d1::task_group_context’ used in nested name specifier
   33 |         tbb::detail::d1::task_group_context::concurrent_wait |
      |                                              ^~~~~~~~~~~~~~~
....../USD/pxr/base/work/dispatcher.cpp:34:46: error: incomplete type ‘tbb::detail::d1::task_group_context’ used in nested name specifier
   34 |         tbb::detail::d1::task_group_context::default_traits)
      |                                              ^~~~~~~~~~~~~~
....../USD/pxr/base/work/dispatcher.h: In member function ‘pxrInternal_v0_22__pxrReserved__::WorkDispatcher::_InvokerTask<typename std::remove_reference<_Tp>::type>& pxrInternal_v0_22__pxrReserved__::WorkDispatcher::_MakeInvokerTask(Fn&&)’:
....../USD/pxr/base/work/dispatcher.h:163:22: error: ‘_rootTask’ was not declared in this scope
  163 |         return *new( _rootTask->allocate_additional_child_of(*_rootTask) )
      |                      ^~~~~~~~~
....../USD/pxr/base/work/dispatcher.h: In member function ‘void pxrInternal_v0_22__pxrReserved__::WorkDispatcher::Run(Callable&&)’:
....../USD/pxr/base/work/dispatcher.h:106:9: error: ‘_rootTask’ was not declared in this scope
  106 |         _rootTask->spawn(_MakeInvokerTask(std::forward<Callable>(c)));
      |         ^~~~~~~~~
....../USD/pxr/base/work/dispatcher.cpp:41:5: error: ‘_rootTask’ was not declared in this scope
   41 |     _rootTask = new(tbb::v1::task::allocate_root(_context)) tbb::empty_task;
      |     ^~~~~~~~~
....../USD/pxr/base/work/dispatcher.cpp:41:36: error: ‘allocate_root’ is not a member of ‘tbb::v1::task’
   41 |     _rootTask = new(tbb::v1::task::allocate_root(_context)) tbb::empty_task;
      |                                    ^~~~~~~~~~~~~
....../USD/pxr/base/work/dispatcher.cpp:41:66: error: ‘empty_task’ in namespace ‘tbb’ does not name a type
   41 |     _rootTask = new(tbb::v1::task::allocate_root(_context)) tbb::empty_task;
      |                                                                  ^~~~~~~~~~
....../USD/pxr/base/work/dispatcher.cpp: In destructor ‘pxrInternal_v0_22__pxrReserved__::WorkDispatcher::~WorkDispatcher()’:
....../USD/pxr/base/work/dispatcher.cpp:48:20: error: ‘destroy’ is not a member of ‘tbb::v1::task’; did you mean ‘tbb::detail::d1::destroy’?
   48 |     tbb::v1::task::destroy(*_rootTask);
      |                    ^~~~~~~
In file included from /opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/enumerable_thread_specific.h:29,
                 from /opt/intel/oneapi/tbb/latest/include/tbb/enumerable_thread_specific.h:17,
                 from ....../USD/pxr/base/tf/diagnosticMgr.h:46,
                 from ....../USD/pxr/base/tf/errorMark.h:30,
                 from ....../USD/pxr/base/work/dispatcher.h:33,
                 from ....../USD/pxr/base/work/dispatcher.cpp:26:
/opt/intel/oneapi/tbb/2021.7.1/include/oneapi/tbb/profiling.h:68:63: note: ‘tbb::detail::d1::destroy’ declared here
   68 |     enum notify_type {prepare=0, cancel, acquired, releasing, destroy};
      |                                                               ^~~~~~~
....../USD/pxr/base/work/dispatcher.h: In member function ‘pxrInternal_v0_22__pxrReserved__::WorkDispatcher::_InvokerTask<typename std::remove_reference<_Tp>::type>& pxrInternal_v0_22__pxrReserved__::WorkDispatcher::_MakeInvokerTask(Fn&&)’:
....../USD/pxr/base/work/dispatcher.h:163:22: error: ‘_rootTask’ was not declared in this scope
  163 |         return *new( _rootTask->allocate_additional_child_of(*_rootTask) )
      |                      ^~~~~~~~~
....../USD/pxr/base/work/dispatcher.cpp:48:29: error: ‘_rootTask’ was not declared in this scope
   48 |     tbb::v1::task::destroy(*_rootTask);
      |                             ^~~~~~~~~
....../USD/pxr/base/work/dispatcher.cpp: In member function ‘void pxrInternal_v0_22__pxrReserved__::WorkDispatcher::Wait()’:
....../USD/pxr/base/work/dispatcher.cpp:55:5: error: ‘_rootTask’ was not declared in this scope
   55 |     _rootTask->wait_for_all();
      |     ^~~~~~~~~
make[2]: *** [pxr/base/work/CMakeFiles/work.dir/build.make:222: pxr/base/work/CMakeFiles/work.dir/utils.cpp.o] Error 1
make[2]: *** [pxr/base/work/CMakeFiles/work.dir/build.make:138: pxr/base/work/CMakeFiles/work.dir/detachedTask.cpp.o] Error 1
make[2]: *** [pxr/base/work/CMakeFiles/work.dir/build.make:152: pxr/base/work/CMakeFiles/work.dir/dispatcher.cpp.o] Error 1
make[2]: Leaving directory '....../USD/build'
make[1]: *** [CMakeFiles/Makefile2:4209: pxr/base/work/CMakeFiles/work.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make[2]: Leaving directory '....../USD/build'
make  -f pxr/imaging/pxOsd/CMakeFiles/pxOsd.dir/build.make pxr/imaging/pxOsd/CMakeFiles/pxOsd.dir/build
make[2]: Entering directory '....../USD/build'
make[2]: Nothing to be done for 'pxr/imaging/pxOsd/CMakeFiles/pxOsd.dir/build'.
make[2]: Leaving directory '....../USD/build'
[ 44%] Built target pxOsd
make[2]: Leaving directory '....../USD/build'
make  -f pxr/base/vt/CMakeFiles/_vt.dir/build.make pxr/base/vt/CMakeFiles/_vt.dir/build
make[2]: Entering directory '....../USD/build'
make[2]: Nothing to be done for 'pxr/base/vt/CMakeFiles/_vt.dir/build'.
make[2]: Leaving directory '....../USD/build'
[ 44%] Built target _vt
make[2]: Leaving directory '....../USD/build'
make  -f pxr/base/gf/CMakeFiles/_gf.dir/build.make pxr/base/gf/CMakeFiles/_gf.dir/build
make[2]: Entering directory '....../USD/build'
make[2]: Nothing to be done for 'pxr/base/gf/CMakeFiles/_gf.dir/build'.
make[2]: Leaving directory '....../USD/build'
[ 45%] Built target _gf
make[1]: Leaving directory '....../USD/build'
make: *** [Makefile:149: all] Error 2
```

**Comment by brechtvl** [comment]
I have a branch here that supports optionally building with oneTBB, and passes the tests on my Linux machine:
https://github.com/brechtvl/USD/commits/onetbb

I can submit this as multiple PRs, but not sure how granular is most convenient. Perhaps one PR with the build system, dispatcher and thread limit changes (since you can't test them individually), and another PR with the various small changes?

**Comment by spiffmon** [comment]
Thank you, @brechtvl - this sounds great!  As to factoring, the first PR you suggested sounds right.  For the rest of the “small” changes, switching to std from tbb objects/APIs, etc, it would be great to get one PR per affected module.  This helps our correctness and performance automation be efficient about tracking regressions, and allows us to not need to revert a big change in the case where it would take time to address an unacceptable regression.  So we would land all of the smaller PR’s one at a time, first, before the build/Work change, since they should all be backwards compatible.

Just to set expectations, we are currently a bit over-booked, but are allotting time in the Fall to address OneTBB compatibility, so that’s when we’ll likely get to reviewing and incorporating these PR’s, and your work should make the process go much faster, so thank you again!

**Comment by mwestphal** [comment]
Please note we are waiting on this at F3D project in order to add USD support in our binary package, thanks a lot for working on this.

https://github.com/f3d-app/f3d/pull/819

**Comment by Meakk** [comment]
Thank you @brechtvl! I tried your fork and it's working fine.  
For those interested, I rebased @brechtvl's fork on `v23.05` and kept their commits and merged commit https://github.com/PixarAnimationStudios/USD/commit/610c6c9e0b8bcc7cdbcfeb294b9c9e0456c5e8f3.  
Here is the [formatted patch](https://gist.github.com/Meakk/c1d5e25c0e4d3dfffb5f776b2f93663b) that one can use to build USD 23.05 with oneTBB.

**Comment by e4lam** [comment]
I'm not familiar with the USD coding style but is it expected that all methods defined inside a class be explicitly marked as inline? It's rather superfluous since any method defined inside a class is already inline.

**Comment by spiffmon** [comment]
Ugh.  Our internal coding conventions (latest revision is still in a draft state, which is part of the reason we have not published them) is silent on that, and it should not be.  I think that **is** the pattern we want, @e4lam , however, we are not entirely consistent, e.g. gf/range.temlplate.h:SetEmpty(), though that is the only inconsistency I noted (but there could be others).

**Comment by brechtvl** [comment]
The addition of `inline` in `pxr/usd/sdf/path.h` is not in my pull requests, it seems to be part of another commit that was included in that formatted patch.

**Comment by alexmyczko** [comment]
is 2021.8.0 also fine enough to be supported? not sure if it's related to https://github.com/PixarAnimationStudios/USD/issues/1650

**Comment by e4lam** [comment]
@brechtvl Yeah, I only realized that later. I think it might have got caught into @Meakk 's formatted patch because it seems like that code in `path.h` was subsequently reverted.

@spiffmon Ok sure. Just curious, what's the motivation for using `inline` on methods defined inside classes?

**Comment by spiffmon** [comment]
There is no justifiable reason, @e4lam ... the offending code is I think more than twenty years old and just has not been scrutinized before now.

**Comment by e4lam** [comment]
@spiffmon Sorry, I thought you meant that what you _want_ is to add the `inline` keyword? Or did I misunderstood you on the other direction?

**Comment by Meakk** [comment]
Is there any update regarding this issue?

**Comment by spiffmon** [comment]
It is queued up for the next few months, thanks.

**Comment by mwestphal** [comment]
Any updates on this ? It doesnt seem to be part of usd 24 sadly.

**Comment by spiffmon** [comment]
It's IP and we hope to land it soon... it threw some wrenches into our larger codebase that took some time to iron out.

**Comment by spiffmon** [comment]
Closing this as we now support 2021.9, officially (since last fall).  Thanks for your patience!
